scipy/optimize/tnc/_tnc_module.pyx

```diff
@@ -0,0 +1,205 @@
+
+    g_data = <float64_t *>np.PyArray_DATA(gx)
+    memcpy(g, g_data, n * sizeof(double))
+
+    return 0
+
+
+cdef void callback_function(double x[], void *state) except *:
+    cdef:
+        pytnc_state *py_state
+        int n
+        double *x_data
+
+    py_state = <pytnc_state *>state
+    n = py_state.n
+
+    # ensures we're working on a copy of the data, just in case user function
+    # mutates it
+    try:
+        xcopy = np.empty(n, dtype=float)
+        x_data = <float64_t *>np.PyArray_DATA(xcopy)
+        memcpy(x_data, x, sizeof(double) * n)
+    except MemoryError as exc:
+        # couldn't allocate memory
+        py_state.failed = 1
+        raise exc
+
+    val = (<object>py_state.py_callback)(xcopy)
+    # TODO examine val to see if we should halt?
+
+
+def tnc_minimize(func_and_grad,
+                   np.ndarray[np.float64_t] x0,
+                   np.ndarray[np.float64_t] low,
+                   np.ndarray[np.float64_t] up,
+                   np.ndarray[np.float64_t] scale,
+                   np.ndarray[np.float64_t] offset,
+                   int messages,
+                   int maxCGit,
+                   int maxfun,
+                   double eta,
+                   double stepmx,
+                   double accuracy,
+                   double fmin,
+                   double ftol,
+                   double xtol,
+                   double pgtol,
+                   double rescale,
+                   callback=None):
+
+    cdef:
+        pytnc_state py_state
+        int n
+        int rc
+        double f = np.inf
+        int nfeval = 0
+        int niter = 0
+        double *x_data
+        double *g_data
+        double *low_data = NULL
+        double *up_data = NULL
+        double *scale_data = NULL
+        double *offset_data = NULL
+        tnc_callback *callback_functionP = NULL
+
+    # initialise structure, use memset?
+    py_state.py_callback = NULL
+    py_state.failed = 0
+
+    if not callable(func_and_grad):
+        raise TypeError("tnc: function must be callable")
+
+    if callback is not None:
+        if not callable(callback):
+            raise TypeError("tnc: callback must be callable or None.")
+        py_state.py_callback = <void*> callback
+        callback_functionP = callback_function
+
+    n = x0.size
+    py_state.n = n
+
+    n3 = np.size(scale)
+    if n3:
+        scale_data = <float64_t *>np.PyArray_DATA(scale)
+
+    n1 = np.size(low)
+    if n1:
+        low_data = <float64_t *>np.PyArray_DATA(low)
+
+    n2 = np.size(up)
+    if n2:
+        up_data = <float64_t *>np.PyArray_DATA(up)
+
+    n4 = np.size(offset)
+    if n4:
+        offset_data = <float64_t *>np.PyArray_DATA(offset)
+
+    if (n1 != n2 or
+        n != n1 or
+        (scale_data != NULL and n != n3) or
+        (offset_data != NULL and n != n4)):
+        raise ValueError("tnc: vector sizes must be equal")
+
+    x = np.copy(x0, order="C")
+    g = np.zeros_like(x, dtype=float)
```

Comment:
Should dtype be `np.float64` just to make it match the pointer case explicitly? - I don't know how strictly specified the size of `float` is (both now and in the future)